kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/MessageFetchContext.java

```diff
@@ -401,8 +405,11 @@ private void readMessagesInternal(KafkaHeaderAndRequest fetch,
                                             ((Integer) THROTTLE_TIME_MS.defaultValue),
                                             ((FetchRequest) fetch.getRequest()).metadata().sessionId()),
                                     () -> {
-                                        responseValues.entrySet().forEach(topicPartitionListEntry -> {
-                                            topicPartitionListEntry.getValue().forEach(Entry::release);
+                                        // release the batched ByteBuf if necessary
+                                        decodeResults.forEach(decodeResult -> {
+                                            if (decodeResult.getReleasedByteBuf() != null) {
+                                                decodeResult.getReleasedByteBuf().release();
+                                            }
```

Comment:
Maybe we should add a `release()` method to `DecodeResult` instead of exposing the internal fields by getters? For example,

```java
// DecodeResult.java
public void release() {
    if (releasedByteBuf != null) {
        releasedByteBuf.release();
    }
}
```

Then here we could just write

```java
decodeResults.forEach(DecodeResult::release);
```

It's only my advice for better code design.

Reply:
And then we could remove the `@Data` annotation as well.